lsp-types/src/Language/LSP/Types/Uri.hs

```diff
@@ -191,20 +209,36 @@ instance Hashable NormalizedFilePath where
   hashWithSalt salt (NormalizedFilePath uri _) = hashWithSalt salt uri
 
 instance IsString NormalizedFilePath where
+    fromString :: String -> NormalizedFilePath
     fromString = toNormalizedFilePath
 
 toNormalizedFilePath :: FilePath -> NormalizedFilePath
-toNormalizedFilePath fp = normalizedFilePath nuri nfp
+toNormalizedFilePath fp = NormalizedFilePath nuri . encodeFilePath $ nfp
   where
-      nfp = FP.normalise fp
-      nuri = internalNormalizedFilePathToUri nfp
+    nfp = FP.normalise fp
+    nuri = internalNormalizedFilePathToUri nfp
 
-fromNormalizedFilePath :: NormalizedFilePath -> FilePath
-fromNormalizedFilePath (NormalizedFilePath _ fp) = fp
+-- | Extracts 'FilePath' from 'NormalizedFilePath'.
+-- The function is total. The 'HasCallStack' constraint is added for debugging purpose only.
```

Comment:
This is a little confusing as it immediately calls error! I get the argument: the invariant of NFP is that it contains a correctly encoded string so this should never fail.

However, I also bet we call this a lot. What would happen if we also cached the decoded filepath in `NormalizedFilePath`? Would it blow out memory usage?

Reply:
Yes, memory usage is the point of using `ShortByteString` or `OsPath` instead of `FilePath`. Despite the encoding/decoding, **both** CPU time and memory usage decreased in benchmarks!

I saw most use sites of `toNormalizedFilePath` just extracts the `FilePath` and pass it to some IO function, so the CPU is not the bottleneck.